### PR TITLE
[Fix #5074] Add Layout/EmptyLinesAroundArguments cop

### LINKED ISSUE #5074: Add new `Layout/EmptyLinesAroundArguments` Cop
While applying a fix for issue #5052 in PR #5066 , it was discovered that the autocorrection for `Style/BracesArondHashParameters` is a little aggressive. When I applied changes to stop it from removing legitimate `\n`s, [it created snippets similar to the example below](https://github.com/bbatsov/rubocop/pull/5066/files?diff=split#diff-0e7dbdd6e99e0defc2bab2cb544b42daR207).

Adding this `EmptyLine` Cop would resolve this scenario and still allow `Style/BracesArondHashParameters` autocorrection to be improved to cleaner results.  The cleaner results I refer to our found in [this example](https://github.com/bbatsov/rubocop/pull/5066/files?diff=split#diff-0e7dbdd6e99e0defc2bab2cb544b42daR683), [this example](https://github.com/bbatsov/rubocop/pull/5066/files?diff=split#diff-4ee3a03eec811b54b5ea49c57363755dR141) and [this example](https://github.com/bbatsov/rubocop/pull/5066/files?diff=split#diff-0e7dbdd6e99e0defc2bab2cb544b42daR203).

--------

## Desired behavior

For this code snippet register a `Stlye/EmptyLinesAroundArguments` offense.

```ruby
foo(

  bar: 1

)
```

Then, autocorrect that snippet to:

```ruby
foo(
  bar: 1
)
```

## Actual behavior

Above snippet registers no offense.

## Steps to reproduce the problem

See above.

## RuboCop version

```
0.51.0 (using Parser 2.4.0.0, running on ruby 2.3.3 x86_64-darwin15)
```

### PR DESCRIPTION
**Pull request by garettarrowood**
Adds `Layout/EmptyLinesAroundArguments` cop.

Ready for review.  Open to suggestions in regards to memoization comments.

-----------------

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`. 
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/layout/empty_lines_around_arguments.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        def autocorrect(node)
+          lambda do |corrector|
+            extra_lines(node) do |range|
+              corrector.remove(range)
+            end
+          end
+        end
+
+        def empty_lines(node)
+          @empty_lines ||= begin
```
Comment (Drenmi) [inline lib/rubocop/cop/layout/empty_lines_around_arguments.rb]: I'm not sure, but I think memoization might mess with the correction loop in some cases.

Comment (garettarrowood) [reply lib/rubocop/cop/layout/empty_lines_around_arguments.rb]: I just tried removing it. And for some reason `rake interal_investigation` then returns offenses:

```
1058 files inspected, 25 offenses detected
RuboCop failed!
```

I got the above results consistently 3 times in a row, only on 3 files - `spec/rubocop/cop/layout/extra_spacing_spec.rb`, `spec/rubocop/cop/layout/indent_array_spec.rb`, & `spec/rubocop/cop/offense_spec.rb`.

Adding the memoization back lets the investigation consistently pass.

Any ideas? I'll see if I can figure it out, but it doesn't make much sense at first inspection.

Comment (garettarrowood) [reply lib/rubocop/cop/layout/empty_lines_around_arguments.rb]: Possibly my cache is corrupt somehow?

File: lib/rubocop/cop/layout/empty_lines_around_arguments.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+module RuboCop
+  module Cop
+    module Layout
+      # This cops checks if empty lines exist around the arguments
+      # of a method invocation.
+      #
+      # @example
+      #   # bad
+      #   do_something(
```
Comment (bbatsov) [inline lib/rubocop/cop/layout/empty_lines_around_arguments.rb]: I guess you can a few more of the examples here - namely something with more arguments.

**Comment by bbatsov** [comment]
The cop seems good to me overall, apart from my tiny comment on documentation.

**Comment by bbatsov** [comment]
👍

**Pull request merged**